api/kroxylicious-filter-api/src/main/java/io/kroxylicious/proxy/filter/FilterInvokers.java

```diff
@@ -29,36 +32,62 @@ private FilterInvokers() {
      * @param filter the Filter to create an invoker for
      * @return the invoker
      */
-    public static FilterInvoker from(KrpcFilter filter) {
+    static List<FilterAndInvoker> from(KrpcFilter filter) {
+        List<FilterAndInvoker> filterInvokers = invokersForFilter(filter);
         // all invokers are wrapped in safe invoker so that clients can safely call onRequest/onResponse
         // even if the invoker isn't interested in that message.
-        return new SafeInvoker(invokerForFilter(filter));
+        return wrapAllInSafeInvoker(filterInvokers).toList();
     }
 
-    private static FilterInvoker invokerForFilter(KrpcFilter filter) {
+    private static List<FilterAndInvoker> invokersForFilter(KrpcFilter filter) {
+        boolean isCompositeFilter = filter instanceof CompositeFilter;
         boolean isResponseFilter = filter instanceof ResponseFilter;
         boolean isRequestFilter = filter instanceof RequestFilter;
         boolean isAnySpecificFilterInterface = SpecificFilterArrayInvoker.implementsAnySpecificFilterInterface(filter);
-        if (isAnySpecificFilterInterface && (isRequestFilter || isResponseFilter)) {
-            throw unsupportedFilterInstance(filter, "Cannot mix specific message filter interfaces and [RequestFilter|ResponseFilter] interfaces");
-        }
-        if (!isRequestFilter && !isResponseFilter && !isAnySpecificFilterInterface) {
-            throw unsupportedFilterInstance(filter, "KrpcFilter must implement ResponseFilter, RequestFilter or any combination of specific message Filter interfaces");
+        validateFilter(filter, isResponseFilter, isRequestFilter, isAnySpecificFilterInterface, isCompositeFilter);
+        if (isCompositeFilter) {
+            List<KrpcFilter> composedFilters = ((CompositeFilter) filter).getFilters();
+            return composedFilters.stream().flatMap(composedFilter -> invokersForFilter(composedFilter).stream()).toList();
```

Comment:
With this we enable nested CompositeFilters, wonder if we should add a configurable max depth

Reply:
Have added a depth limit of 2, so you can write:
```CompositeFilter([CompositeFilter([Filter1, Filter2]), CompositeFilter([Filter3, Filter4])])` and end up with `[Filter1, Filter2, Filter3, Filter4]` in the chain, but no deeper.